### PR TITLE
Gateway base contract unit test

### PR DESCRIPTION
**Pull request by sarvesh-ost**
1. Unit test cases for GatewayBase.

2. Added mock gateway base, mock gatewaylib, and mock core.

#368

### CODE REVIEW

File: test/gateway/gateway_base/bounty_change.js [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+    it('should be able to re-propose bounty change', async function () {
+
+      let proposedBounty = new BN(50);
+      let currentBounty = new BN(100);
+
+      await proposeBountyChange(gatewayBaseInstance, proposedBounty, organisation, currentBounty);
+
+       proposedBounty = new BN(150);
+       currentBounty = new BN(100);
```
Comment (schemar) [inline test/gateway/gateway_base/bounty_change.js]: same as above, right?

File: test/gateway/gateway_base/bounty_change.js [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+      await proposeBountyChange(gatewayBaseInstance, proposedBounty, organisation, currentBounty);
+    });
+
+    it('only organization should be able to propose bounty change', async function () {
```
Comment (schemar) [inline test/gateway/gateway_base/bounty_change.js]: ideally, these should generally read like a sentence in JS tests. Starting with the `it`. Like `it should not accept a bounty change from another address than the organization` or something like that.

File: test/gateway/gateway_base/bounty_change.js [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+      let proposedBounty = new BN(50);
+      await  Utils.expectThrow(gatewayBaseInstance.initiateBountyAmountChange.call(proposedBounty, {from: accounts[1]}));
```
Comment (schemar) [inline test/gateway/gateway_base/bounty_change.js]: it is very long, you could split it into multiple lines.
```js
Utils.expectThrow(
    gatewayBaseInstance.initiateBountyAmountChange.call(
        proposedBounty,
        {from: accounts[1]}
    )
);
```

File: test/gateway/gateway_base/get_nonce.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    });
+
+    it('should return nonce incremented by 1 if stake process is initiated', async function () {
+
+     //todo implement this when stake unit tests are done
+    });
+
+    it('should return nonce incremented by 1 if linking process is initiated', async function () {
+
+      //todo implement this when liking unit tests are done
```
Comment (schemar) [inline test/gateway/gateway_base/get_nonce.js]: *linking

File: test/gateway/gateway_base/prove_gateway.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        response.receipt.status,
+        1,
+        "Receipt status is unsuccessful"
+      );
+
+      let eventData = response.logs;
+      await Utils.validateEvents(eventData, expectedEvent);
+    });
+
+    it('should be not prove  with if already proven', async function () {
```
Comment (schemar) [inline test/gateway/gateway_base/prove_gateway.js]: please correct sentence.

File: test/gateway/gateway_base/prove_gateway.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+      assert.equal(
+        response.receipt.status,
+        1,
+        "Receipt status is unsuccessful"
+      );
+
+      let eventData = response.logs;
+
+      await Utils.validateEvents(eventData, expectedEvent);
```
Comment (schemar) [inline test/gateway/gateway_base/prove_gateway.js]: how does this assert that it is not proven again when it actually checks for the event (compare title)? 😕

File: test/gateway/gateway_base/prove_gateway.js [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    });
+
+    it('should fail if rlp encodedAccount is not passed', async function () {
+
+      await Utils.expectThrow(gatewayBaseInstance.proveGateway(1, "0x", rlpParentNodes));
+
+    });
+
+    it('should fail if rlp rlpParentNodes is not passed', async function () {
```
Comment (schemar) [inline test/gateway/gateway_base/prove_gateway.js]: please check this sentence.

File: test/lib/utils.js [hunk 8]
```diff
@@ -20,165 +20,243 @@
...
+  /*
+   *  Basic Ethereum checks
+   */
 
 /// @dev Compare to null address
-module.exports.isNullAddress = function (address) {
-  Assert.strictEqual(typeof address, 'string', `address must be of type 'string'`);
-  return (address == NullAddress);
-}
+  isNullAddress: function (address) {
```
Comment (schemar) [inline test/lib/utils.js]: no `function` keyword here (compare to others) :)
(also some functions below)

File: contracts/test/MockGatewayBase.sol [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+        // for a block height
+        emit GatewayProven(
+            remoteGateway,
+            _blockHeight,
+            storageRoot,
+            false
+        );
+
+        return true;
+    }
```
[the remark this answers is not in the document]
Comment (sarvesh-ost) [reply contracts/test/MockGatewayBase.sol]: This contract uses an external library GatewayLib. To get stub data from GatewayLib, I have created MockGatewayBase and MockGatewayLib.

File: test/gateway/gateway_base/activate_gateway.js [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        messageBus,
+        bounty,
+        organisation
+      );
+
+    });
+
+    it('should deactivate if activated', async function () {
+
+      assert((await gatewayBaseInstance.deactivateGateway.call({from: organisation})));
```
[the remark this answers is not in the document]
Comment (sarvesh-ost) [reply test/gateway/gateway_base/activate_gateway.js]: call on deactivateGateway will return solidity return values which is bool in this case.

File: contracts/test/MockGatewayBase.sol [hunk 11]
```diff
@@ -17,10 +16,13 @@ contract MockGatewayBase is GatewayBase {
...
-    {
-
-    }
+        GatewayBase(
+            _core,
+            _messageBus,
+            _bounty,
+            _organisation
+        )
+    {}
```
[the remark this answers is not in the document]
Comment (deepesh-kn) [reply contracts/test/MockGatewayBase.sol]: I prefer `{}`

File: contracts/test/MockGatewayLib.sol [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+
+    /**
+     *	@notice Mock Merkle proof verification of account.
+     *
+     *	@param _rlpEncodedAccount rlp encoded data of account.
+     *	@param _rlpParentNodes path from root node to leaf in merkle tree.
+     *	@param _encodedPath encoded path to search account node in merkle tree.
+     *	@param _stateRoot state root for given block height.
+     *
+     *	@return bytes32 Storage path of the variable
```
[the remark this answers is not in the document]
Comment (sarvesh-ost) [reply contracts/test/MockGatewayLib.sol]: Sure, I have seen this pattern in the style guide.

File: test/gateway/gateway_base/activate_gateway.js [hunk 13]
```diff
@@ -64,7 +64,9 @@ contract('GatewayBase.sol', function (accounts) {
 
     it('should activate if deActivated', async function () {
 
-      assert((await gatewayBaseInstance.activateGateway.call({from: organisation})));
+      assert((await gatewayBaseInstance.activateGateway.call({from: organisation})),
+             "Gateway activation failed, activateGateway returned false."
+            );
```
Comment (schemar) [inline test/gateway/gateway_base/activate_gateway.js]: Should be:
```js
      assert(
        (await gatewayBaseInstance.activateGateway.call({from: organisation})),
        "Gateway activation failed, activateGateway returned false.",
      );
```

File: contracts/test/MockGatewayBase.sol [hunk 14]
```diff
@@ -93,7 +93,7 @@ contract MockGatewayBase is GatewayBase {
             return true;
         }
         
-        //on successful proof verification storage root is returned other wise
+        //On successful proof verification storage root is returned other wise
```
Comment (schemar) [inline contracts/test/MockGatewayBase.sol]: Missing space in front of `On`

**Comment by schemar** [comment]
Thank you!! LGTM! @deepesh-kn if you have no further comments you can merge 😊

**Pull request merged**